mon: manage endpoints containing set of mon ip addresses

patrostkowski opened the pull request:
The change creates and manages an Endpoints resource with the current Ceph monitor (mon) IPs, allowing clients to resolve mon IPs via DNS without relying on the rook-ceph-mon-endpoints ConfigMap.

Resolves: https://github.com/rook/rook/issues/14986

<!-- Thank you for contributing to Rook! -->

<!-- STEPS TO FOLLOW:
  1. Add a description of the changes (frequently the same as the commit description)
  2. Enter the issue number next to "Resolves #" below (if there is no tracking issue resolved, **remove that section**)
  3. Review our Contributing documentation at https://rook.io/docs/rook/latest/Contributing/development-flow/
  4. Follow the steps in the checklist below, starting with the **Commit Message Formatting**.
-->

<!-- Uncomment this section with the issue number if an issue is being resolved
**Issue resolved by this Pull Request:**
Resolves #
--->

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

patrostkowski commented on pkg/operator/ceph/cluster/mon/mon.go:
```diff
@@ -1188,7 +1194,57 @@ func (c *Cluster) saveMonConfig() error {
+		}
+	}
+
+	logger.Infof("created mon endpoints %+v", endpoints.Subsets)
```
we can keep it as info level I guess, it will not clutter the output that much as CM method logs the same thing.

patrostkowski commented on pkg/operator/ceph/cluster/mon/mon.go:
```diff
@@ -1188,7 +1194,57 @@ func (c *Cluster) saveMonConfig() error {
 }
 
-func (c *Cluster) persistExpectedMonDaemons() error {
+func (c *Cluster) persistExpectedMonDaemonsAsEndpoints() error {
```
sure, that approach looks more tidy

haslersn commented:
For me, this can wait until 1.17. I might backport it myself for testing.

Madhu-1 commented on pkg/operator/ceph/cluster/mon/mon.go:
```diff
@@ -1189,6 +1191,66 @@ func (c *Cluster) saveMonConfig() error {
+		})
+	}
+
+	endpoints := &v1.Endpoints{
```
v1.Endpoints is deprecated and removed going forward isnt it, cant we use endpointSlice? https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/4974-deprecate-endpoints

patrostkowski replied:
oh I was unaware of this, sounds like a good idea to use EndpointSlice instead

Madhu-1 commented on pkg/operator/ceph/cluster/mon/mon.go:
```diff
@@ -1189,6 +1191,66 @@ func (c *Cluster) saveMonConfig() error {
+
+	err := c.ownerInfo.SetControllerReference(endpoints)
+	if err != nil {
+		return errors.Wrapf(err, "failed to set owner reference mon endpoint %q", endpoints.Name)
```
```suggestion
return errors.Wrapf(err, "failed to set controller reference mon endpoint %q", endpoints.Name)
```

patrostkowski commented on pkg/operator/ceph/cluster/mon/mon.go:
```diff
@@ -1189,6 +1192,73 @@ func (c *Cluster) saveMonConfig() error {
+		AddressType: "IPv4",
+	}
+
+	logger.Infof("creating mon endpointslice %+v", endpointSlice)
```
ah thats some log leftovers, removing as it is not required.

patrostkowski commented:
@travisn as @haslersn said I am also ok to wait for 1.17

haslersn reviewed (changes requested):
I backported the code to v1.16.5 in order to test it on our rook v1.16.5 cluster. I observed that the `rook-ceph-active-mons` EndpointSlice gets created and DNS resolution (and mounting via that DNS name) works. Though I haven't tested whether the EndpointSlice gets properly updated on mon failover.

Madhu-1 commented on pkg/operator/ceph/cluster/mon/mon.go:
```diff
@@ -1189,6 +1192,71 @@ func (c *Cluster) saveMonConfig() error {
+			},
+		},
+		Ports:       nil,
+		AddressType: "IPv4",
```
If the mon pod is running on IPV6 only will this also work without any issue?

Madhu-1 commented on pkg/operator/ceph/cluster/mon/mon.go:
```diff
@@ -1189,6 +1192,71 @@ func (c *Cluster) saveMonConfig() error {
+	monAddresses := make([]string, 0, len(monitors))
+
+	for _, mon := range monitors {
+		ip := strings.Split(mon.Endpoint, ":")[0]
```
dont this cause issue for ipv6 IP's?

haslersn replied:
Good catch. The code currently doesn't support IPv6 – see my other comment about the fact that we need separate EndpointSlice objects. And then, we need to add the `mon.Endpoint` to the correct EndpointSlice object, depending on whether it's an IPv4 or IPv6 `mon.Endpoint`.

patrostkowski replied:
yeah good catch everyone. Give me a few days to test this out with IPv6 and dual stack. I will post the changes very soon... busy week :D

patrostkowski commented:
@travisn @Madhu-1 @haslersn I tested the changes in both IPv6-only and dual-stack environments. Below are my findings. I have some questions because I find it hard to understand how ipFamily is used and how c.ClusterInfo.AllMonitors() works. I would really appreciate your help in finding the right way to handle this.

1. IPv6 Only (ipFamily=IPv6, dualStack=false)

```
apiVersion: ceph.rook.io/v1
kind: CephCluster
metadata:
  name: rook-ceph
  namespace: rook-ceph
spec:
  network:
    ipFamily: IPv6
    dualStack: false
```

```
│ 2025-03-18 21:30:53.019998 D | op-mon: updating config map rook-ceph-mon-endpoints that already exists                                                                                   │
│ 2025-03-18 21:30:53.220275 I | op-mon: saved mon endpoints to config map map[csi-cluster-config-json:[{"clusterID":"rook-ceph","monitors":["[2a01:4ff:1f0:cbe2::800:4587]:6789","[2a01:4 │
│ ff:1f0:cbe2::800:c7f3]:6789","[2a01:4ff:1f0:cbe2::800:b6f5]:6789"],"cephFS":{"netNamespaceFilePath":"","subvolumeGroup":"","radosNamespace":"","kernelMountOptions":"","fuseMountOptions │
│ ":""},"rbd":{"netNamespaceFilePath":"","radosNamespace":"","mirrorDaemonCount":0},"nfs":{"netNamespaceFilePath":""},"readAffinity":{"enabled":false,"crushLocationLabels":null},"namespa │
│ ce":""}] data:b=[2a01:4ff:1f0:cbe2::800:4587]:6789,c=[2a01:4ff:1f0:cbe2::800:c7f3]:6789,a=[2a01:4ff:1f0:cbe2::800:b6f5]:6789 externalMons: mapping:{"node":{"a":{"Name":"fedora","Hostna │
│ me":"fedora","Address":"fddd:2bc9:985:749:dc80:eff:fe0c:d501"},"b":{"Name":"fedora","Hostname":"fedora","Address":"fddd:2bc9:985:749:dc80:eff:fe0c:d501"},"c":{"Name":"fedora","Hostname │
│ ":"fedora","Address":"fddd:2bc9:985:749:dc80:eff:fe0c:d501"}}} maxMonId:2 outOfQuorum:]                                                                                                  │
│ 2025-03-18 21:30:53.220298 D | op-mon: check if should create IPv4 endpointslice: false (IPFamily=IPv6, DualStack=false)                                                                 │
│ 2025-03-18 21:30:53.220300 D | op-mon: check if should create IPv6 endpointslice: true (IPFamily=IPv6, DualStack=false)                                                                  │
│ 2025-03-18 21:30:53.223186 D | op-mon: updating endpointslice resource rook-ceph-active-mons-ipv6 that already exists                                                                    │
│ 2025-03-18 21:30:53.225462 I | op-mon: created IPv6 mon endpointslice [{Addresses:[2a01:4ff:1f0:cbe2::800:c7f3 2a01:4ff:1f0:cbe2::800:b6f5 2a01:4ff:1f0:cbe2::800:4587] Conditions:{Read │
│ y:<nil> Serving:<nil> Terminating:<nil>} Hostname:<nil> TargetRef:nil DeprecatedTopology:map[] NodeName:<nil> Zone:<nil> Hints:nil}]
```

```
$ k run --rm -it --image heywoodlh/dnsutils dns -- sh
If you don't see a command prompt, try pressing enter.
/ # nslookup rook-ceph-active-mons.rook-ceph.svc.cluster.local
Server:		2a01:4ff:1f0:cbe2::800:a
Address:	2a01:4ff:1f0:cbe2::800:a#53

Name:	rook-ceph-active-mons.rook-ceph.svc.cluster.local
Address: 2a01:4ff:1f0:cbe2::800:4587
Name:	rook-ceph-active-mons.rook-ceph.svc.cluster.local
Address: 2a01:4ff:1f0:cbe2::800:b6f5
Name:	rook-ceph-active-mons.rook-ceph.svc.cluster.local
Address: 2a01:4ff:1f0:cbe2::800:c7f3
```

IPv6 EndpointSlice is created with the correct addresses and the headless service correctly resolves the mon addresses.

2. Dual-stack (ipFamily=IPv4, dualStack=true)

```
apiVersion: ceph.rook.io/v1
kind: CephCluster
metadata:
  name: rook-ceph
  namespace: rook-ceph
spec:
  network:
    ipFamily: IPv4
    dualStack: true
```

```
│ 2025-03-18 22:00:37.786051 I | op-mon: saved mon endpoints to config map map[csi-cluster-config-json:[{"clusterID":"rook-ceph","monitors":["10.43.199.231:6789","10.43.77.53:6789","10.4 │
│ 3.248.184:6789"],"cephFS":{"netNamespaceFilePath":"","subvolumeGroup":"","radosNamespace":"","kernelMountOptions":"","fuseMountOptions":""},"rbd":{"netNamespaceFilePath":"","radosNames │
│ pace":"","mirrorDaemonCount":0},"nfs":{"netNamespaceFilePath":""},"readAffinity":{"enabled":false,"crushLocationLabels":null},"namespace":""}] data:b=10.43.77.53:6789,c=10.43.248.184:6 │
│ 789,a=10.43.199.231:6789 externalMons: mapping:{"node":{"a":{"Name":"fedora","Hostname":"fedora","Address":"192.168.0.128"},"b":{"Name":"fedora","Hostname":"fedora","Address":"192.168. │
│ 0.128"},"c":{"Name":"fedora","Hostname":"fedora","Address":"192.168.0.128"}}} maxMonId:2 outOfQuorum:]                                                                                   │
│ 2025-03-18 22:00:37.786089 D | op-mon: check if should create IPv4 endpointslice: true (IPFamily=IPv4, DualStack=true)                                                                   │
│ 2025-03-18 22:00:37.786091 D | op-mon: check if should create IPv6 endpointslice: true (IPFamily=IPv4, DualStack=true)                                                                   │
│ 2025-03-18 22:00:37.789214 D | op-mon: updating endpointslice resource rook-ceph-active-mons-ipv4 that already exists                                                                    │
│ 2025-03-18 22:00:37.790964 I | op-mon: created IPv4 mon endpointslice [{Addresses:[10.43.77.53 10.43.248.184 10.43.199.231] Conditions:{Ready:<nil> Serving:<nil> Terminating:<nil>} Hos │
│ tname:<nil> TargetRef:nil DeprecatedTopology:map[] NodeName:<nil> Zone:<nil> Hints:nil}]                                                                                                 │
│ 2025-03-18 22:00:38.788340 D | op-mon: monConfig: &{ResourceName:rook-ceph-mon-a DaemonName:a PublicIP:10.43.199.231 Port:6789 Zone: NodeName:fedora DataPathMap:0x4000ec3950 UseHostNet │
│ work:false}
```

```
$ k run --rm -it --image heywoodlh/dnsutils dns -- sh
If you don't see a command prompt, try pressing enter.
/ # nslookup rook-ceph-active-mons.rook-ceph.svc.cluster.local
Server:		fd02::a
Address:	fd02::a#53

Name:	rook-ceph-active-mons.rook-ceph.svc.cluster.local
Address: 10.43.77.53
Name:	rook-ceph-active-mons.rook-ceph.svc.cluster.local
Address: 10.43.248.184
Name:	rook-ceph-active-mons.rook-ceph.svc.cluster.local
Address: 10.43.199.231
```

`c.ClusterInfo.AllMonitors()` returns only IPv4 addresses so only IPv4 EndpointSlice is created and the headless service correctly resolves IPv4 addresses. However, no IPv6 EndpointSlice is created because of that. Not sure if thats because we do not care about IPv6 when ipFamily is IPv4?

2.Dual-stack (ipFamily=IPv6, dualStack=true)

```
apiVersion: ceph.rook.io/v1
kind: CephCluster
metadata:
  name: rook-ceph
  namespace: rook-ceph
spec:
  network:
    ipFamily: IPv6
    dualStack: true
```

```
│ 2025-03-18 22:06:15.094234 I | op-mon: saved mon endpoints to config map map[csi-cluster-config-json:[{"clusterID":"rook-ceph","monitors":["10.43.210.202:6789","10.43.91.218:6789","10. │
│ 43.108.66:6789"],"cephFS":{"netNamespaceFilePath":"","subvolumeGroup":"","radosNamespace":"","kernelMountOptions":"","fuseMountOptions":""},"rbd":{"netNamespaceFilePath":"","radosNames │
│ pace":"","mirrorDaemonCount":0},"nfs":{"netNamespaceFilePath":""},"readAffinity":{"enabled":false,"crushLocationLabels":null},"namespace":""}] data:c=10.43.108.66:6789,a=10.43.210.202: │
│ 6789,b=10.43.91.218:6789 externalMons: mapping:{"node":{"a":{"Name":"fedora","Hostname":"fedora","Address":"192.168.0.128"},"b":{"Name":"fedora","Hostname":"fedora","Address":"192.168. │
│ 0.128"},"c":{"Name":"fedora","Hostname":"fedora","Address":"192.168.0.128"}}} maxMonId:2 outOfQuorum:]                                                                                   │
│ 2025-03-18 22:06:15.094255 D | op-mon: check if should create IPv4 endpointslice: true (IPFamily=IPv6, DualStack=true)                                                                   │
│ 2025-03-18 22:06:15.094258 D | op-mon: check if should create IPv6 endpointslice: true (IPFamily=IPv6, DualStack=true)                                                                   │
│ 2025-03-18 22:06:15.094544 D | ceph-spec: object "rook-ceph-mon-endpoints" matched on update                                                                                             │
│ 2025-03-18 22:06:15.094555 D | ceph-spec: do not reconcile on configmap "rook-ceph-mon-endpoints"                                                                                        │
│ 2025-03-18 22:06:15.096823 D | op-mon: updating endpointslice resource rook-ceph-active-mons-ipv4 that already exists                                                                    │
│ 2025-03-18 22:06:15.099094 I | op-mon: created IPv4 mon endpointslice [{Addresses:[10.43.210.202 10.43.91.218 10.43.108.66] Conditions:{Ready:<nil> Serving:<nil> Terminating:<nil>} Hos │
│ tname:<nil> TargetRef:nil DeprecatedTopology:map[] NodeName:<nil> Zone:<nil> Hints:nil}]
```

```
$ k run --rm -it --image heywoodlh/dnsutils dns -- sh
If you don't see a command prompt, try pressing enter.
/ # nslookup rook-ceph-active-mons.rook-ceph.svc.cluster.local
Server:		fd02::a
Address:	fd02::a#53

Name:	rook-ceph-active-mons.rook-ceph.svc.cluster.local
Address: 10.43.210.202
Name:	rook-ceph-active-mons.rook-ceph.svc.cluster.local
Address: 10.43.91.218
Name:	rook-ceph-active-mons.rook-ceph.svc.cluster.local
Address: 10.43.108.66
```

Only IPv4 monitor addresses are returned, even though ipFamily=IPv6 and dualStack=true. I suspect `c.ClusterInfo.AllMonitors()` is only fetching status.podIP, which defaults to IPv4. Shouldn't it be using status.podIPs to fetch both IPv4 and IPv6?

4. And the strange issue, where I do not specify any network settings:

```
apiVersion: ceph.rook.io/v1
kind: CephCluster
metadata:
  name: rook-ceph
  namespace: rook-ceph
spec:
  network: {}
```

```
│ 2025-03-18 20:46:26.181206 I | op-mon: saved mon endpoints to config map map[csi-cluster-config-json:[{"clusterID":"rook-ceph","monitors":["[2a01:4ff:1f0:cbe2::800:cf32]:6789","[2a01:4 │
│ ff:1f0:cbe2::800:3638]:6789"],"cephFS":{"netNamespaceFilePath":"","subvolumeGroup":"","radosNamespace":"","kernelMountOptions":"","fuseMountOptions":""},"rbd":{"netNamespaceFilePath":" │
│ ","radosNamespace":"","mirrorDaemonCount":0},"nfs":{"netNamespaceFilePath":""},"readAffinity":{"enabled":false,"crushLocationLabels":null},"namespace":""}] data:a=[2a01:4ff:1f0:cbe2::8 │
│ 00:cf32]:6789,b=[2a01:4ff:1f0:cbe2::800:3638]:6789 externalMons: mapping:{"node":{"a":{"Name":"fedora","Hostname":"fedora","Address":"fddd:2bc9:985:749:dc80:eff:fe0c:d501"},"b":{"Name" │
│ :"fedora","Hostname":"fedora","Address":"fddd:2bc9:985:749:dc80:eff:fe0c:d501"},"c":{"Name":"fedora","Hostname":"fedora","Address":"fddd:2bc9:985:749:dc80:eff:fe0c:d501"}}} maxMonId:1  │
│ outOfQuorum:]
│ 2025-03-18 20:46:26.011103 D | op-mon: check if should create IPv4 endpointslice: false (IPFamily=, DualStack=false)                                                                 │
│ 2025-03-18 20:46:27.011231 D | op-mon: check if should create IPv6 endpointslice: true (IPFamily=, DualStack=false)  
```

In this case no EndpointSlice gets created. I evaluate the creation of the resources by:

```
createIPv4 := c.spec.Network.IPFamily == cephv1.IPv4 || c.spec.Network.DualStack
createIPv6 := c.spec.Network.IPFamily == cephv1.IPv6 || c.spec.Network.DualStack
```
Because IPFamily is empty, the values for `createIPv4` and `createIPv6` default to false. Debug logs show that IPFamily default to empty value. Should IPFamily be defaulted to IPv4 somewhere in the code?

haslersn commented:
@patrostkowski  I'm not a Rook maintainer, so we have to ask them.

- Maybe when dualStack is true, ipFamily is ignored and Rook always uses IPv4 mon endpoints? That would explain your points 2 and 3.
- Maybe an empty ipFamily is equivalent to ipFamily=IPv4? that would explain your point 4.

Another note: Do you also delete the EndpointSlices when these conditions change, i.e., when dualStack/ipFamily of a CephCluster changes? Maybe it's easier to always create and reconcile both IPv4 and IPv6 EndpointSlices, even in single stack clusters. Then we don't have to implement the deletion logic.

> `c.ClusterInfo.AllMonitors()` returns only IPv4 addresses so only IPv4 EndpointSlice is created and the headless service correctly resolves IPv4 addresses. However, no IPv6 EndpointSlice is created because of that.

I assume an EndpointSlice can contain an empty list of endpoints, so when the general conditions for creating an EndpointSlice for some address family are met, you should IMO create and reconcile it even if there are no endpoints for that specific address family.

patrostkowski commented:
> Do you also delete the EndpointSlices when these conditions change

@haslersn not sure if its even possible to change the network fields without breaking everything. And for the address updates I went with following approach: get whatever `c.ClusterInfo.AllMonitors()`returns, determine if addresses are IPv4 or IPv6, enforce IPv4 if `ipFamily` is empty, create/update EndpointSlices for both IPv4 and IPv6 if there are any addresses.

> I assume an EndpointSlice can contain an empty list of endpoints, so when the general conditions for creating an EndpointSlice for some address family are met, you should IMO create and reconcile it even if there are no endpoints for that specific address family. Otherwise it would again require some more deletion logic.

@haslersn  No need for deletion logic, any addresses returned by `c.ClusterInfo.AllMonitors()` will be transformed into correct format and the operator will update the existing EndpointSlices.

@sp98 I have pushed the changes, let me know if that covers all scenarios.

haslersn commented:
> @haslersn not sure if its even possible to change the network fields without breaking everything.

Maybe other things break, I don't know. But that doesn't mean we should introduce yet another feature that breaks.

haslersn commented on pkg/operator/ceph/cluster/mon/mon.go:
```diff
@@ -1189,6 +1193,124 @@ func (c *Cluster) saveMonConfig() error {
+		}
+	}
+
+	createIPv4 := c.spec.Network.IPFamily == cephv1.IPv4 || c.spec.Network.DualStack
```
Why not `!= cephv1.IPv6`, then you don't need the special case below.

```suggestion
	createIPv4 := c.spec.Network.IPFamily != cephv1.IPv6 || c.spec.Network.DualStack
```

haslersn commented on pkg/operator/ceph/cluster/mon/mon.go:
```diff
@@ -1189,6 +1193,124 @@ func (c *Cluster) saveMonConfig() error {
+		createIPv4 = true
+	}
+
+	if createIPv4 && len(monAddressesIPv4) > 0 {
+		if err := c.createOrUpdateEndpointSlice(monAddressesIPv4, "IPv4"); err != nil {
+			return err
+		}
+	}
+
+	if createIPv6 && len(monAddressesIPv6) > 0 {
+		if err := c.createOrUpdateEndpointSlice(monAddressesIPv6, "IPv6"); err != nil {
+			return err
+		}
+	}
```
Currently, EndpointSlice updates are missed in the following cases:

- If one (or more) IPv4 mon endpoints existed in the past, but now only IPv6 mon endpoints exist, the former IPv4 endpoints are not removed from the IPv4 EndpointSlice. Solution: simply drop the check `len(monAddressesIPv4) > 0` and let `createOrUpdateEndpointSlice` write an empty list.
- If createIPv4 was true in the past and now is false, the IPv4 EndpointSlice is leaked.
- The same two cases symmetrically with IPv6.

BlaineEXE commented on pkg/operator/ceph/cluster/mon/mon.go:
```diff
@@ -1189,6 +1193,124 @@ func (c *Cluster) saveMonConfig() error {
+	for _, mon := range monitors {
+		ipPort := mon.Endpoint
+
+		var ip string
+		if strings.HasPrefix(ipPort, "[") {
+			ip = ipPort[1:strings.Index(ipPort, "]")]
+		} else {
+			ip = strings.Split(ipPort, ":")[0]
+		}
```
Is `net.SplitHostPort()` not viable here? https://pkg.go.dev/net#SplitHostPort

patrostkowski commented:
@haslersn sorry for not answering your comments, but I found some interesting case which may confirm that creating IPv4/IPv6 EndpointSlices was wrong all along.

so I set up dev envs again (only IPv4, only IPv6, dual-stack), I deployed operator with basic CephCluster in a few configurations as previously and it looks like:

- in IPv4 and dual-stack configurations, the addresses that `c.ClusterInfo.AllMonitors()` return are IPv4 **only**
- even when `dualStack` is true, the returned addresses from `c.ClusterInfo.AllMonitors()` in scenarios above are IPv4 anyway
- in dual-stack configuration, the addresses returned by `c.ClusterInfo.AllMonitors()` are Service addresses anyway
- dual-stack configuration, all Services created by Rook when `dualStack` is true are `SingleStack` anyway

@travisn @sp98 @BlaineEXE could you confirm, that we are just not expecting `c.ClusterInfo.AllMonitors()` to return both IPv4 and IPv6 addresses for all Mons? If we do not expect that, then there is no need to differentiate the EndpointSlice by IP type, as there is no such mechanism for saving the mon endpoints to the ConfigMap.

patrostkowski commented:
> I see it as sufficient to have a single endpoint slice with the results of `AllMonitors()`

@travisn thank you for the detailed explanation! I have added this reasoning as a comment in `persistExpectedMonDaemonsAsEndpointSlice` to clarify why we treat monitor IPs as single-stack identities, even in dual-stack clusters.

I also removed the use of `c.spec.Network.IPFamily` in favor of directly parsing the IP addresses to determine their family. In previous tests, `IPFamily` was behaving inconsistently, so relying on the actual address content proved to be more robust.

haslersn commented:
@travisn 

When transitioning from a DualStack cluster to an IPv6 cluster, I assume that one would recreate one mon at a time. During migration, the new mons have IPv6 addresses while the old mons have IPv4 addresses.

> If both ipv4 and ipv6 addresses were generated for the same mon, I don't expect that to work since it would be two identities for the same mon. So I see it as sufficient to have a single endpoint slice with the results of `AllMonitors()`.

I don't agree with the conclusion. Even if a single mon only has a single IP address, there could be a mix of families, e.g., in the scenario described above.

patrostkowski commented:
@haslersn I agree, this behaviour with `SingleStack` as default Services is weird and Rook should definitely respect the configurations with given `.spec.network.ipFamily`. Maybe we could file another issue and describe the concerns there?

patrostkowski commented:
@travisn @haslersn ok, I have adjusted the logic to parse IPs from `AllMonitors()` and create EndpointSlices based on returned IP types 😄

haslersn commented on pkg/operator/ceph/cluster/mon/mon.go:
```diff
@@ -1189,6 +1195,105 @@ func (c *Cluster) saveMonConfig() error {
+
+func (c *Cluster) createEndpointSliceForAddresses(addresses []string, addressType discoveryv1.AddressType) error {
+	if len(addresses) == 0 {
+		logger.Debugf("no %s addresses found, skipping endpointslice creation", addressType)
+		return nil
```
At this point you should delete the EndpointSlice if it exists.

haslersn commented on pkg/operator/ceph/cluster/mon/mon.go:
```diff
@@ -1188,6 +1194,113 @@ func (c *Cluster) saveMonConfig() error {
+func (c *Cluster) persistExpectedMonDaemonsAsEndpointSlice() error {
+	monitors := c.ClusterInfo.AllMonitors()
+	if len(monitors) == 0 {
+		logger.Debug("no mon addresses found, skipping endpointslice resource reconciliation")
```
```suggestion
		// Theoretically, we could now go ahead with the normal code path to
		// delete (both IPv4 and IPv6) EndpointSlices, but 0 mons is certainly
		// an error state, so it's better to do nothing destructive right now.
		logger.Debug("no mon addresses found, skipping endpointslice resource reconciliation")
```

haslersn commented:
I backported the commit 28b4d9478c417ce6eba4087d29789514eabebdbf to v1.16.5 again and it works on my CephCluster with no custom network settings.

haslersn commented on pkg/operator/ceph/cluster/mon/mon.go:
```diff
@@ -1188,6 +1194,116 @@ func (c *Cluster) saveMonConfig() error {
+	if len(addresses) == 0 {
+		logger.Debugf("no %s addresses found, deleting existing %q endpointslice if exists", addressType, sliceName)
+		if err := client.Delete(c.ClusterInfo.Context, sliceName, metav1.DeleteOptions{}); err != nil {
+			if !kerrors.IsNotFound(err) {
```
I'm currently working on improving the tests for this MR. My new tests cought the following bug:

```suggestion
			if kerrors.IsNotFound(err) {
```

patrostkowski replied:
hmm whats your test procedure? I tried checking this case by creating empty endpointslice with different IPFamily and it worked.

haslersn replied:
Actually, my tests didn't catch that bug, but I saw the bug in the logs, just like @travisn.

patrostkowski replied:
ouch indeed, it should not be negated this way

haslersn commented on pkg/operator/ceph/cluster/mon/mon.go:
```diff
@@ -1188,6 +1194,116 @@ func (c *Cluster) saveMonConfig() error {
+			if !kerrors.IsNotFound(err) {
+				logger.Debugf("endpointslice %q not found, nothing to delete", sliceName)
+			} else {
+				logger.Errorf("failed to delete endpointslice %q: %v", sliceName, err)
```
It's just that the if condition is wrongly negated.

haslersn commented:
@travisn 

> just one question... The port does not need to be set? Maybe not since you're observing it working, but just wanted to confirm when comparing it to other endpoint slices created automatically for the services.

The port(s) should be irrelevant for headless services, since kube-proxy has nothing to do with them. Only kube-dns, and kube-dns and doesn't care about ports.

However, I didn't find any official statement in the docs, stating that `EndpointSlice.ports` is allowed to be `null`.

patrostkowski commented:
> However, I didn't find any official statement in the docs, stating that `EndpointSlice.ports` is allowed to be `null`.

@travisn @haslersn from [official](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.27/#endpointslice-v1-discovery-k8s-io) docs: `When ports is empty, it indicates that there are no defined ports. When a port is defined with a nil port value, it indicates "all ports"` so maybe its better to attach empty list to indicate that we really dont care about ports at all.

haslersn commented:
> When a port is defined with a nil port value

By this they mean

```yaml
ports:
  - port: null
```

but we have:

```yaml
ports: null # This is probably equivalent to `ports: []`
```

patrostkowski commented:
> This is probably equivalent to `ports: []`

yup, API server does not fail when creating such EndpointSlice, but as @travisn said I will just attach the ports for consistency

haslersn commented:
There was a compile error in my tests, now fixed. I edited the link to point to my branch where the fixed commit is now pushed.

patrostkowski commented:
@haslersn The tests look great! I have added them to the suite and also included port validation.

haslersn commented on pkg/operator/ceph/cluster/mon/mon_test.go:
```diff
@@ -278,24 +280,210 @@ func TestPersistMons(t *testing.T) {
+	assert.Equal(t, discoveryv1.AddressTypeIPv4, epSliceIPv4.AddressType)
+	assert.Len(t, epSliceIPv4.Endpoints, 1)
+	assert.ElementsMatch(t, ipv4Addresses, epSliceIPv4.Endpoints[0].Addresses)
+	assert.Len(t, epSliceIPv4.Ports, len(expectedPortsNoRequireMsgr2))
+	assertEndpointSlicePorts(t, epSliceIPv4.Ports, expectedPortsNoRequireMsgr2)
```
Why check for the length here, if we check the whole array anyway? `assertEndpointSlicePorts()` already contains a length check.

Anyway, I suggest to simply use `assert.ElementsMatch()` and get rid of the `assertEndpointSlicePorts()` method, like so: https://github.com/haslersn/rook/commit/7be0673075863ea915fd9a690246eb4acdcca4ab

haslersn commented on pkg/operator/ceph/cluster/mon/mon_test.go:
```diff
@@ -278,24 +280,210 @@ func TestPersistMons(t *testing.T) {
+	assert.ElementsMatch(t, ipv6Addresses, epSliceIPv6.Endpoints[0].Addresses)
+	assertEndpointSlicePorts(t, epSliceIPv6.Ports, expectedPortsNoRequireMsgr2)
+
+	// RequireMsgr2=true
```
This is code duplication. I suggest to extract the testing logic into a function that gets called with different settings, like so: https://github.com/haslersn/rook/commit/5a2e99784547b76bc53e1dbc7550435bb04f3d34

patrostkowski replied:
yup, the deduplication approach is way better for sure. I pulled your changes, thank you!

haslersn reviewed (changes requested):
Apart from the suggested changes below, I observed that you also omitted the checks for non-existence of certain EndpointSlices. I suggest to add them, like so: https://github.com/haslersn/rook/commit/71733f88f5b65a0097338130cc20d478ff3687f2

My full list of requested changes can be found here: https://github.com/haslersn/rook/compare/d9eed0b994e9364b660e76736d4c5d1e8e845c2a...haslersn:rook:mon-active-ep-14986-suggestion2

parth-gr commented:
@patrostkowski @travisn 
With the above fix i see issues while setting up the rook ceph cluser,
```
| ceph-cluster-controller: failed to reconcile CephCluster "openshift-storage/ocs-storagecluster-cephcluster". failed to reconcile cluster "ocs-storagecluster-cephcluster": failed to configure local ceph cluster: failed to create cluster: failed to start ceph monitors: failed to initialize ceph cluster info: failed to save mons: failed to persist expected mon daemons: failed to persist expected mons in EndpointSlice: failed to create IPv4 endpointslice: endpointslices.discovery.k8s.io "rook-ceph-active-mons-ipv4" is forbidden: endpoint address 172.30.109.184 is not allowed
```

Probably we need to revert this fix, or there is any change in setting we need to perform?

haslersn commented:
I have found this issue – https://github.com/zalando/postgres-operator/issues/1702 – which suggests that manually configuring Endpoints/EndpointSlices is not possible in OpenShift. Is that true? What's the alternative for OpenShift?

parth-gr commented:
> I have found this issue – [zalando/postgres-operator#1702](https://github.com/zalando/postgres-operator/issues/1702) – which suggests that manually configuring Endpoints/EndpointSlices is not possible in OpenShift. Is that true? What's the alternative for OpenShift?

Added the fix https://github.com/rook/rook/pull/15645, thanks!

patrostkowski commented:
oh didn’t know about this imitation, thanks @parth-gr
